compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -682,23 +682,39 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             };
 
             let encl_item_id = self.tcx.hir().get_parent_item(expr.hir_id);
-            let encl_item = self.tcx.hir().expect_item(encl_item_id);
 
-            if let hir::ItemKind::Fn(..) = encl_item.kind {
-                // We are inside a function body, so reporting "return statement
-                // outside of function body" needs an explanation.
+            if self.tcx.hir().maybe_body_owned_by(encl_item_id).is_some() {
```

Comment:
Isn't this check redundant with the following `if let`?

Reply:
No, because a trait method does not have a body if it does not have a default implementation, but it can still contain a misplaced return expression, e.g.:
```rust
fn foo(a: [(); return]);
```

Reply:
What about matching on `hir::TraitItemKind::Fn(_, hir::TraitFn::Provided(_))`?

Reply:
Yes, that should also work. Fixed.